---
title: "Pr√©ambules"
output:
  learnr::tutorial:
    theme: default
    highlight: default
    fig_caption: yes
    self_contained: true
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 4
    number_sections: false
    progressive: true
  html_document:
    theme: cosmo
    fig_caption: yes
    self_contained: yes
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 3
    number_section: true
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
runtime: shiny_prerendered
---

<!--  
Here the parameters about the documents.
https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf 
--> 


```{css, echo=FALSE}

```

<script language="JavaScript" type="text/javascript">
          
          function sizeTbl2(h,i) {
          var tbl = document.getElementById("section-" + i);
          tbl.style.display = h;
          }

</script>



<style>
.exo {
  border-radius: 5px;
  margin-top: 5px;
  margin-bottom: 5px;
  padding-top: 5px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  background-color: #fcede3;
  color: rgb(51, 51,153);
}
.tips {
       padding-top: 5px;
       padding-bottom: 5px;
       padding-left: 5px;
       padding-right: 5px;
       border: 1px dashed #2f6fab;
       background-color: #EEFFEE;
}
.solution {
            margin-top: 5px;
            margin-bottom: 5px;
            padding-top: 5px;
            padding-bottom: 5px;
            padding-left: 5px;
            padding-right: 5px;
            border: 1px dashed #FFFFFF;
            background-color: #EEEEFF;
            color: #0000BB;
            font-size: 11px;
            hidden=true;
}
</style>


```{r setup, include=FALSE}
library(learnr)
library(gradethis)
knitr::opts_chunk$set(echo=TRUE, eval=TRUE, cache=FALSE, message=FALSE, warning=FALSE, comment="")
gradethis::gradethis_setup()
```


## A few words about R

The advent of cutting-edge technologies in the field of **high-throughput biological analysis techniques** has led to the exponential production of data. As a result, it has become commonplace for biologists to manipulate data sets comprising **thousands, millions or even billions of values**. In this context, it is essential to have flexible software capable of **automating the processes of collecting, processing and statistically analyzing these data**.

R is a data processing and statistical analysis software package originally developed by **Robert Gentleman and Ross Ihaka**. It runs on a variety of platforms including Unix, Linux, Windows and Macintosh, and enjoys free software status. The freedom inherent in its use implies constant evolution, made possible by an active **community of developers who share their improvements** on a regular basis.

R's usefulness extends to a vast **community of scientists** from diverse backgrounds, such as biologists, statisticians, mathematicians and geographers. This ecosystem fosters the exchange of tools developed and encourages the convergence of disciplines towards **collaborative projects**. This dynamic is undoubtedly the driving force behind the tool's success.

R stands out for its flexibility and **relative ease of use**. Unlike compiled languages such as C/C++, Fortran or Pascal, which require the creation of a complete program before execution, R is an **interpreted language**. In other words, command lines are executed interactively** by the machine. However, R's approach can sometimes confuse novice users, requiring a certain amount of experience to master its workings.

The initial investment in learning R is **quickly rewarded**. In fact, a **multitude of tools is made available in the form of libraries**, commonly known as *packages*. Each *package* contains a range of high-level functions dedicated to specific tasks. This richness enables biologists to exploit R's potential for the statistical analysis of data from both "classical" **biological methods and "high-throughput" methods, such as the transcriptome, epigenome, proteome or interactome**.

Last but not least, R is equipped with a **sophisticated range of graphical tools** that enable the creation of fully customizable outputs. Analyzed data can be exported in various formats such as .jpg, .png, .bmp, .ps, .pdf, .html and .txt, enhancing the ability to share and present results in a varied and attractive way.

## Object creation

### Assignment operators

R is an object-based language (*vector, matrix, data.frame, list, factor,*...). We generally use the '<-' (arrow) operator to create objects. To begin with, let's create a simple "vector" object (a kind of array) that will contain a numerical value. Here, the code "x <- 15" could read "the variable x takes the value 15" or "we assign 15 to x".

```{r assign0, exercise=TRUE}
x <- 15
print(x)
```

The arrow assignment operator '<-' can be substituted by the '=' operator (this syntax is often used by R language beginners). However, it is **always preferable to use** '<-', which indicates that a value (*15*) is stored in a variable named *(x)*. This is the recommendation of Hadley Wickham, developer of numerous function libraries, in his [*tidyverse style guide*](https://style.tidyverse.org/) (*cf* section 2.7).

```{r assign1, exercise=TRUE, exercise.setup = "assign0"}
x = 15
print(x)
```

The following form (*i.e* the inverted arrow) is accepted. However, it will be encountered **extremely rarely** and is not recommended (again, by stylistic convention).

```{r assign0b, exercise=TRUE}
15 -> x
print(x)
```

<div class="exo">
- Write a code to store 3 in a variable x, 5 in y and then the sum (operator *+*) of x and y in z. Print the contents of the 3 variables.

```{r assign2, exercise=TRUE, exercise.setup = "assign1"}

```


```{r assign2-solution, eval=TRUE, echo=FALSE}
x <- 3
y <- 5
z <- x + y
print(x)
print(y)
print(z)
```

```{r assign2-check, eval=TRUE, echo=FALSE}
gradethis::grade_result_strict(
  pass_if(~ x == 3),
  pass_if(~ y == 5),
  pass_if(~ z == 8)
)
```
</div>

### How to name objects

There are several recommendations for naming objects.

- It is generally recommended to use explicit, lower-case variable names. 
- Variable names should be in English. All other language elements are in English.  Moreover, in the scientific world, you're likely to be working with non-French speakers...
- We'll follow the recommendations of [Hadley Wickman](http://adv-r.had.co.nz/Style.html) and use an "underscore" character (_) to separate words in a variable name. 
- Try to keep names concise and meaningful. Here are a few examples.


```{r assign3, exercise=TRUE}
age <- 20
average_size <- 130
nb_iteration <- 100
nb_genes <- 20e3
```

**In this tutorial, we take a few liberties by naming the variables x, y or z. This allows you to write quickly and to the point. This is not recommended for your production code.

## Code elements

### Comments

Code may be commented using the "#" character. Any command following this character will not be interpreted. This allows you to annotate your code (that is explain how it works for yourself and others).

```{r assign3b, exercise=TRUE}
z <- 10
# The code below is not interpreted. The z variable is therefore unchanged.
# z <- 56
print(z)
```

<div class="exo">
- Add the following comment to the code below: "The value 10 is assigned to z".

```{r comment, exercise=TRUE, exercise.setup = "assign1"}
z <- 10
```


```{r comment-solution, eval=TRUE, echo=FALSE}
# The value 10 is assigned to z
z <-  10
```

```{r comment-check, eval=TRUE, echo=FALSE}
grade_code()
```

</div>

### Instruction separation

Instructions can be separated by a carriage return or by the **';'** character.

```{r assign4, exercise=TRUE}
x <- 12; y <- 13
cat(x) # <=> print(x)
cat(y) # <=> print(y)
```

<div class="exo">
- Add, on the same line of code: y <- 5

```{r semicolumn, exercise=TRUE, exercise.setup = "assign1"}
z <- 10
```


```{r semicolumn-solution, eval=TRUE, echo=FALSE}
z <-  10 ;  y <- 5
```

```{r semicolumn-check, eval=TRUE, echo=FALSE}
grade_code(allow_partial_matching=FALSE)
```

</div>

## Manage objects in memory

### Listing and destroying objects

Created objects are stored in the computer's RAM. At this stage, they are not stored on disk. If you quit the program without saving, they will be destroyed. 

To find out which objects are available (*i.e* have been created at a given time), you can **list** the objects in memory with the **ls()** function. 

```{r assign5, exercise=TRUE}
# Objects are created
x <- 1; y <- 2; z <- 3
# List objects in memory
ls()
```

Objects can be destroyed using the **rm()** function (**remove**).

```{r assign5b, exercise=TRUE, exercise.setup="assign5"}
# Destroy x 
rm(x)
ls()
```

### Predefined variables

There are a number of predefined variables  in the R environment. These include `LETTERS` (upper-case alphabet), `letters` (lower-case alphabet), `pi` (...). We'll use them in our code according to our needs.

```{r letters, exercise=TRUE}
letters
```
```{r letters_2, exercise=TRUE}
LETTERS
```
```{r pi, exercise=TRUE}
LETTERS
```


## Functions

### Functions ? 

The **objects** *x*, *y* and *z* created earlier are vectors for the R language. Here, they each contain a numerical value. We've also used the **functions** **ls()** and **rm()**. In R, you can call functions to query objects and perform actions on them. Functions take the following form:

```
functionname(arg1= a, arg2= b,...)

# Above arg1 and arg2 (...) correspond to argument names
# a and b correspond to the objects/values you wish to pass to the 
# 'functionname' function to perform a task using them.

```

For example, you can use the `sum()` function to add vectors. However, we can substitute the '+' operator (which, in fact, is similar to a hidden function).


```{r function0, exercise=TRUE }
# Both lead to the same
# result
sum(17, 15)
17 + 15
```

<div class="exo">

- Use the `ls()` function with argument `pattern='a'` to list all objects in memory with an 'a' in their name. Store the result in a variable named 'z'.

```{r lsexo, exercise=TRUE}
abc <- 3 
bla <- 1 
foo <- 2
bar <- 43
toto <- 56
tutu <- 23
var <- 7
```

```{r lsexo-solution}
abc <- 3 
bla <- 1 
foo <- 2
bar <- 43
toto <- 56
tutu <- 23
var <- 7
z <- ls(pattern='a')
```


```{r lsexo-check}
gradethis::grade_result_strict(
  pass_if(~ all(c("abc", "bar", "bla", "var") %in% z))
)
```

</div>


### Examples of function usage

The `c()` (**combine**) function is used to combine values (e.g. numerical) in a vector (*vector*). The `length()` function lets you find out the size of a vector. 

**NB:** below 1:8 produces a vector containing all integers from 1 to 8. 

```{r function1, exercise=TRUE}
# Integers from 1 to 8
x <- 1:8
# Combine -50, x (i.e. 1 to 8) and 70 
# into a new vector y
y <- c(-50, x, 70)
print(y)

# The length of y (i.e the number of elements in y)
print(length(y))
```

The `is()` and `class()` functions are used to find out the type of an object. Knowing the type of an object is important for knowing which functions can be applied to it. 

```{r function2, exercise=TRUE}
y <- c(2.567, 3.987, 4, 10.34566)
is(y)
class(y)
```

<div class="exo">

- Store the type of x (`class()`) in z.

```{r is, exercise=TRUE}
x <- matrix()
```

```{r is-solution}
x <- matrix()
z <- class(x)
```


```{r is-check}
gradethis::grade_result_strict(
  pass_if(~ z[1] == "matrix")
)
```

</div>

Below is an example of how to use the `round()` function to round numerical values. 

```{r fonction3, exercise=TRUE}
y <- c(2.567, 3.987, 4, 10.34566)
round(y)
```

The `round()` function has several arguments: *x* and *digits* (which controls the number of digits behind the decimal point).


```{r}
args(round)
```

```{r fonction3b, exercise=TRUE}
y <- c(2.567, 3.987, 4, 10.34566)
round(x=y, digits = 2)
```

Note that **argument names can be abbreviated** (note that abbreviation can be ambiguous when there are many arguments with the same prefix).

```{r fonction4, exercise=TRUE}
y <- c(2.567, 3.987, 4, 10.34566)
round(x=y, d = 2)
```

**Argument names can also be omitted**. However, they must be in the expected order.

```{r fonction4b, exercise=TRUE}
y <- c(2.567, 3.987, 4, 10.34566)
round(y, 2)
```

## Getting help in R

### The help() function

Help on a given function can be obtained using the `help()` function. Help can also be called up using the "?" character. Help sheets about functions provide information on input arguments, returned values and usage through examples.

The various help fields are as follows:

- **Description:** provides information on the function's role.
- **Usage:** describes the function and its default parameters.
- **Arguments:** lists the arguments to be passed to this function.
- **Value:** the type of object(s) returned by the function.
- **See Also:** other similar or complementary functions.
- **Examples:** some examples of how to use this function}.

<div class="exo">
Ask for help for the `mean()` function:

```{r function6, exercise=TRUE, exercise.setup = "function1"}

```

```{r function6-solution}
?mean
# or
help(mean)
```


<div class="exo">

- Look for help on some of the following calculation functions: `min()`, `max()`, `sort()`, `log()`, `log2()`, `log10()`. 

```{r findhelp, exercise=TRUE}

```

```{r findhelp-solution}
?min
?max
?sort
?log
?log2
?log10

```

</div>


### The apropos() function

the function `apropos()` may be very handy, as it returns all functions whose names contain a searched string (e.g. `apropos('Sys')` to search for all functions whose names contain 'Sys'). Often, functions in programming have meaningful naming, which helps a lot. Please note that this function is case-sensitive (*i.e* lowercase/lowercase).

<div class="exo">
Find all functions whose name contains 'read'. Please note that 'read' must be enclosed in quotation marks.

<div class="tips">
**NB**: In computing, we use the term **'single quote'** to refer to the apostrophe (' ') and **'double quote'** to refer to quotation marks (‚Äù ‚Äù). In the R language, both forms have the same meaning, and are used to **delimit a sequence of characters** called a *character string* (here **'read' or ‚Äúread‚Äù**). 
</div>

```{r apropos, exercise=TRUE }

```

```{r apropos-solution}
apropos("read")
```

</div>

## Exercises

### Exercise 1

Answer the following questions:

<div class="exo">
```{r exoq1, echo=FALSE }
learnr::question("It's advisable to use the '=' operator rather than '<-' when assigning ?",
  answer("Yes"),
  answer("No", correct = TRUE),
  allow_retry = TRUE 
)
```

```{r exoq2, echo=FALSE }
learnr::question("When calling a function, if the arguments are named, their order doesn't matter.",
  answer("Yes", correct = TRUE),
  answer("No"),
  allow_retry = TRUE 
)
```

```{r exoq4, echo=FALSE }
learnr::question("When calling a function, if you abbreviate the names of the arguments, take care that the abbreviations are not ambiguous.",
  answer("Yes", correct = TRUE),
  answer("No"),
  allow_retry = TRUE 
)
```

```{r exoq3, echo=FALSE }
learnr::question("The ls() function is used to list files?",
  answer("Yes"),
  answer("No", correct = TRUE),
  allow_retry = TRUE 
)
```

</div>

### Finished

Thank you for following this tutorial.
