---
# theme
# “bootstrap”, “cerulean”, “cosmo”, “darkly”, “flatly”, “journal”, “lumen”, “paper”, 
# “readable”, “sandstone”, “simplex”, “spacelab”, “united”, “yeti”
# highlight: `default`, `tango`, `pygments`, `kate`, `monochrome`, `espresso`, `zenburn`, 
# `haddock`, `breezedark`, `textmate`, `arrow`, or `rstudio` or a file with extension `.theme`.

title: "The ggplot2 library (session 4)"
output:
  learnr::tutorial:
    includes:
      before_body: !expr system.file(file.path("tutorials", "style.html"),package="rtrainer")
    theme: default
    highlight: default
    fig_caption: yes
    self_contained: true
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 4
    number_sections: false
    progressive: true
  html_document:
    theme: cosmo
    fig_caption: yes
    self_contained: yes
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 3
    number_section: true
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
runtime: shiny_prerendered
---

<!--  
Here the parameters about the documents.
https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf 
--> 

```{css, echo=FALSE}

```

<script language="JavaScript" type="text/javascript">
          
          function sizeTbl2(h,i) {
          var tbl = document.getElementById("section-" + i);
          tbl.style.display = h;
          }

</script>



<style>
.exo {
  border-radius: 5px;
  margin-top: 5px;
  margin-bottom: 5px;
  padding-top: 5px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  background-color: #fcede3;
  color: rgb(51, 51,153);
}
.tips {
       padding-top: 5px;
       padding-bottom: 5px;
       padding-left: 5px;
       padding-right: 5px;
       border: 1px dashed #2f6fab;
       background-color: #EEFFEE;
}
.solution {
            margin-top: 5px;
            margin-bottom: 5px;
            padding-top: 5px;
            padding-bottom: 5px;
            padding-left: 5px;
            padding-right: 5px;
            border: 1px dashed #FFFFFF;
            background-color: #EEEEFF;
            color: #0000BB;
            font-size: 11px;
}
</style>


```{r setup, include=FALSE}
library(learnr)
library(gradethis)
library(ggplot2)
knitr::opts_chunk$set(echo=TRUE, eval=TRUE, cache=FALSE, message=FALSE, warning=FALSE, comment="")
gradethis::gradethis_setup()
```

## Facets

### Principle

The `ggplot2` library offers an extremely powerful tool for dividing a plot into panels (facets) based on the levels of specified categorical variables. **Facets allow for data exploration based on a factor** or a given group of factors. For the following example, we will create a matrix containing the results of a **fictitious ELISA test**, where measurements are taken at 2 different times (days) for experiments conducted by four different operators.

```{r read_table, exercise=TRUE }

url <- "https://zenodo.org/record/8210893/files/elisa_artificial.txt"
elisa <- read.table(url, sep="\t", header=TRUE, row.names=1)
head(elisa)

```

This is an artificial dataset from an ELISA test using 96-well plates. Eight ELISA plates (12 columns / 8 rows) were used, as can be verified here.

```{r table, exercise=TRUE, exercise.setup="read_table" }
table(elisa$rows, elisa$columns)
```

These eight plates were produced by 4 experimenters on two different days.

```{r table2, exercise=TRUE, exercise.setup="read_table" }
table(elisa$user, elisa$day)
```

### The facet_wrap() Function

With the `ggplot2` syntax, it becomes very easy to produce histograms corresponding to the intensity of colorations obtained in each well (`value`) for a given experimenter (`user`). In the example below, note the use of the `facet_wrap()` function:

- This function creates **a one-dimensional arrangement of facets**, which can optionally be displayed across multiple rows using the `nrow` and `ncol` arguments.
- The `facets` argument passed to `facet_wrap()` must be a formula (`formula`). In our example, `facets = ~ user` translates to 'create graphical panels based on the value of the `user` variable.'


```{r facet_wrap, exercise=TRUE, exercise.setup="read_table", fig.height=6, fig.width=6}
p <- ggplot(data = elisa, 
            mapping = aes(x=value))
p + geom_histogram() + 
    facet_wrap(facets = ~ user, ncol=2)
```

For exploratory purposes, we can similarly analyze the distributions of the values obtained based on the operator (`user`) and the day (`day`).


```{r facet_wrap2, exercise=TRUE, exercise.setup="read_table", fig.height=8, fig.width=6}
p <- ggplot(data = elisa, 
            mapping = aes(x=value))
p + geom_histogram() + 
    facet_wrap(facets = ~ user + day, ncol=2)
```

<div class="exo">

- Display the density distributions of `value` in facets based on the experimenter and the day.

```{r userday_dens, exercise=TRUE, echo=FALSE, fig.height=8, fig.width=6, exercise.setup="read_table"}
```

```{r userday_dens-solution}
p <- ggplot(data = elisa, 
            mapping = aes(x=value, fill=user))
p + geom_density(color=NA) + 
    facet_wrap(facets = ~ user + day, ncol=2)
```
</div>

### The facet_grid() Function

Since each user performed an ELISA experiment on Monday and Friday, we can choose a **two-dimensional faceted representation** with `facet_grid()` (a grid/matrix of facets). Note that the `facets` argument is set to `user ~ day`, indicating that `user` will be displayed in rows and `day` in columns.


```{r facet_grid, exercise=TRUE, exercise.setup="read_table" }
p <- ggplot(data = elisa, 
            mapping = aes(x=value, fill=user))
p + geom_histogram() + 
    facet_grid(facets = user ~ day) +
    scale_fill_manual(values=c("#1B9E77", "#D95F02", "#7570B3", "#E7298A"))
```

<div class="exo">
- Display horizontal boxplots (using `coord_flip()`) corresponding to the distributions of the variable `value` for each user, creating a facet based on the day.


```{r boxplotfacet_show, echo=FALSE, fig.height=6, fig.width=6}
url <- "https://zenodo.org/record/8210893/files/elisa_artificial.txt"
elisa <- read.table(url, sep="\t", header=TRUE, row.names=1)
p <- ggplot(data = elisa, 
            mapping = aes(x=user, y=value, fill=user))
p + geom_boxplot() + coord_flip() +
    facet_wrap(facets = ~ day, ncol=2)
```

```{r boxplotfacet, exercise=TRUE, echo=FALSE, fig.height=6, fig.width=6, exercise.setup="read_table"}

```

```{r boxplotfacet-solution}
p <- ggplot(data = elisa, 
            mapping = aes(x=user, y=value, fill=user))
p + geom_boxplot() + coord_flip() +
    facet_wrap(facets = ~ day, ncol=2)
```
</div>

## Application Example: **Heatmap**

Based on the numerical data loaded into R, we may want to create a color-coded image (**heatmap**) of the ELISA plates produced by different users.

- The coordinates `x` (`elisa$rows`) and `y` (`elisa$columns`) of the wells in the plate are available.
- The `fill` aesthetic will be mapped to `value`.

We can use *geom_raster()* to represent an ELISA plate and partition the plot based on `user` and `day`.

```{r geom_raster, exercise=TRUE, exercise.setup="read_table", fig.height=8, fig.width=6 }
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, y=rows, fill=value))
p + geom_raster() + 
  facet_grid(facets = user ~ day) 
```

For `geom_raster()`, which represents **the continuous numeric variable `value`**, one of the following functions can be used to control the colors:

- `scale_fill_gradient()`: This function is used to specify the fill colors in a gradual manner in a plot, using a single start color and a single end color.

- `scale_fill_gradient2()`: This function is similar to `scale_fill_gradient()`, but it allows specifying an additional color, which serves as a midpoint or central value in the color scale, creating a two-color gradient.

- `scale_fill_gradientn()`: This function is used to specify a fill gradient with multiple custom colors. You can define the colors you want to use in the color scale based on your data and preferences.

```{r scale_fill_gradientn, exercise=TRUE, exercise.setup="read_table" }
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, y=rows, fill=value))
p + geom_raster() + 
  facet_grid(facets = user ~ day) +
  scale_fill_gradientn(colours = c("#0000BF", "#0000FF", 
                                   "#0080FF", "#00FFFF", 
                                   "#40FFBF", "#80FF80", 
                                   "#BFFF40", "#FFFF00", 
                                   "#FF8000", "#FF0000", 
                                   "#BF0000"))
```

### Ordering Rows/Columns

You may have noticed that the rows are not ideally ordered. We would prefer the order: 'cont', 'a', 'b', 'c', ...

In `ggplot2`, to order factors, you need to create ordinal variables. This can be done, as we saw earlier, using the `ordered=TRUE` argument in the `factor()` function.

<div class="exo">

- Modify the following code so that the columns are ordered.

```{r, echo=FALSE, fig.height=8, fig.width=6}
url <- "https://zenodo.org/record/8210893/files/elisa_artificial.txt"
elisa <- read.table(url, sep="\t", header=TRUE, row.names=1)

elisa$rows <- factor(x = elisa$rows, ordered = T, levels=c('cont', letters[1:7]))
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, 
                          y=rows, 
                          fill=value))
p <- p + geom_raster() + 
          facet_grid(facets = user ~ day) +
          scale_fill_gradientn(colours = c("#0000BF", "#0000FF", 
                                           "#0080FF", "#00FFFF", 
                                           "#40FFBF", "#80FF80", 
                                           "#BFFF40", "#FFFF00", 
                                           "#FF8000", "#FF0000", 
                                           "#BF0000"))
print(p)
```

```{r orderedfactor, exercise=TRUE, exercise.setup="read_table", fig.height=8, fig.width=6}
___
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, y=rows, fill=value))
p + geom_raster() + 
  facet_grid(facets = user ~ day) +
  scale_fill_gradientn(colours = c("#0000BF", "#0000FF", 
                                   "#0080FF", "#00FFFF", 
                                   "#40FFBF", "#80FF80", 
                                   "#BFFF40", "#FFFF00", 
                                   "#FF8000", "#FF0000", 
                                   "#BF0000"))
```

```{r orderedfactor-solution}
elisa$rows <- factor(x = elisa$rows, ordered = T, levels=c('cont', letters[1:7]))
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, 
                          y=rows, 
                          fill=value))
p <- p + geom_raster() + 
          facet_grid(facets = user ~ day) +
          scale_fill_gradientn(colours = c("#0000BF", "#0000FF", 
                                           "#0080FF", "#00FFFF", 
                                           "#40FFBF", "#80FF80", 
                                           "#BFFF40", "#FFFF00", 
                                           "#FF8000", "#FF0000", 
                                           "#BF0000"))
print(p)
```

```{r orderedfactor-check}

gradethis::grade_result_strict(
  pass_if(~ levels(p$data$rows)[1] == "cont"),
  pass_if(~ max(ggplot2::ggplot_build(p)$data[[1]]$group) == 96)
)
```
</div>

## Predefined Graphic Themes

### Introduction to Themes

There are many ways to adjust the overall visual appearance of a plot. As a first step, you can apply a **predefined theme**, which affects various parameters of the plot (fonts, character sizes, axis styles, background color, contrast, etc.). ggplot2 includes around ten built-in themes. The names of these global configuration functions usually start with 'theme_'.


```{r}
apropos("^theme_")
```

For example:

- `theme_gray()`: The signature theme of ggplot2 with a gray background and white grid lines, designed to highlight the data while facilitating comparisons.

```{r theme_setup, echo=FALSE, fig.width=3, fig.height=3}
set.seed(123)
x <- rnorm(40)
y <- x + rnorm(40, sd=0.25)
df <- data.frame(x = x, y = y)
base <- ggplot(df, aes(x, y)) + geom_point()
base + theme_gray() + ggtitle("theme_gray()")
```

- `theme_bw()`: The classic ggplot2 theme with a white background and black grid lines, designed to highlight the data while facilitating comparisons. It may be better suited for presentations displayed using a projector.

```{r theme_setup_1, echo=FALSE, fig.width=3, fig.height=3}
base + theme_bw() + ggtitle("theme_bw()")
```

- `theme_linedraw()`: A theme with only black lines of varying widths on a white background, reminiscent of a line drawing. The goal is similar to that of `theme_bw()`. 

```{r theme_setup_2, echo=FALSE, fig.width=3, fig.height=3}
base + theme_linedraw() + ggtitle("theme_linedraw()")
```

- `theme_void()`: A totally empty theme.

```{r theme_setup_3, echo=FALSE, fig.width=3, fig.height=3}
base + theme_void() + ggtitle("theme_void()")
```

- `theme_minimal()`: un thème totalement épuré.

```{r theme_setup_4, echo=FALSE, fig.width=3, fig.height=3}
base + theme_minimal() + ggtitle("theme_minimal()")
```

- `theme_dark()` dont le nom est très parlant...

```{r theme_setup_5, echo=FALSE, fig.width=3, fig.height=3}
base + theme_dark() + ggtitle("theme_dark()")
```

- `theme_classic()`, `theme_test()`, `theme_dark()`, `theme_light()`...

D'autres thèmes prédéfinis sont disponibles dans la librairie [ggthemes](https://rpubs.com/Mentors_Ubiqum/ggthemes_1).

- Par exemple `theme_excel()` pour les nostalgiques du tableur Microsoft... L'aide indique: "Thème permettant de reproduire l'affreuse monstruosité qu'était l'ancien graphique Excel à fond gris. Ne l'utilisez jamais." :). A noter que vous pouvez aussi bénéficier de l'hideuse palette excel (`scale_colour_excel()`). Un must... :)

```{r theme_setup_6, echo=FALSE, fig.width=3, fig.height=3}
base + ggthemes::theme_excel() + ggtitle("theme_excel()")
```

- Ou encore `theme_wsj()` pour simuler un diagramme du Wall Street Journal...

```{r theme_setup_7, echo=FALSE, fig.width=3, fig.height=3}
base + ggthemes::theme_wsj() + ggtitle("theme_wsj()")
```

### Exercice

<div class="exo">

- Essayez successivement d'ajouter l'un des thèmes suivant au diagramme $p$:  `theme_bw()`, `theme_classic()`, `theme_dark()`, `theme_gray()`, `theme_grey()`, `theme_light()`, `theme_minimal()`, `theme_void()`, `ggthemes::theme_wsj()`, `ggthemes::theme_excel()`, `ggthemes::theme_excel_new()`, `ggthemes::theme_economist()`...


```{r preppenguins, echo=FALSE}
library(palmerpenguins)
data("penguins")
penguins <- na.omit(penguins)

# Create a custom color palette
custom_palette <- c("dodgerblue", "darkorange", "forestgreen")

# Create the colored bubble plot
p <- ggplot(penguins, aes(x = flipper_length_mm, y = bill_length_mm, size = body_mass_g, color = species)) +
    geom_point(alpha = 0.7) +
    scale_color_manual(values = custom_palette) +
    labs(x = "Flipper Length (mm)", 
         y = "Bill Length (mm)", 
         color = "Species", 
         size = "Body Mass (g)") 
```

```{r theme2, exercise=TRUE, exercise.setup="preppenguins" }
p <- p + theme_bw()
print(p)
```

```{r theme2-solution}
p <- p + theme_bw()
print(p)
p <- p + theme_classic()
print(p)
p <- p + theme_dark()
print(p)
p <- p + theme_light()
print(p)
p <- p + theme_minimal()
print(p)
p <- p + ggthemes::theme_wsj()
print(p)
p <- p + ggthemes::theme_excel()
print(p)
p <- p + ggthemes::theme_excel_new()
print(p)
p <- p + ggthemes::theme_economist()
print(p)
#...
```

</div>

## Fine-Tuning Graphs

### The theme() and element_*() Functions

Beyond applying predefined themes (like `theme_minimal()` or `theme_bw()`), you can customize every aspect of a graph according to your needs.

- The `theme()` function offers **maximum flexibility for customizing the appearance of your graphs**. For example, you can **specify the font, font size, and text color for axis titles and legends, or change the background of the graph** to fit a dark or light theme.

- **The `element_*()` functions (notably `element_text()`, `element_line()`, `element_rect()`, `element_blank()`...)** are used **in combination with `theme()`** to control specific elements of the graph.

- The **`element_text()` function is used to define the font, font size, and color of a text element**.

- The **`element_line()` function customizes graph lines, such as line thickness or line type**.

- The **`element_rect()` function customizes box/rectangle-type elements**.

- You can use **`element_blank()` to completely remove certain elements** of the graph if needed.

<div class="exo">

- Place your cursor between the parentheses and press the <TAB> key on your keyboard to view all the arguments of `theme()`. This reveals all the modifiable elements of the graph (and there are many...).


```{r theme, exercise=TRUE }
theme()
```

```{r theme-solution}
# rect
# text
# title
# aspect.ratio
# axis.title
# axis.title.x
# axis.title.x.top
# axis.title.x.bottom
# axis.title.y
# axis.title.y.left
# axis.title.y.right
# axis.text
# axis.text.x
# axis.text.x.top
# axis.text.x.bottom
# axis.text.y
# axis.text.y.left
# axis.text.y.right
# axis.ticks
# axis.ticks.x
# axis.ticks.x.top
# axis.ticks.x.bottom
# axis.ticks.y
# axis.ticks.y.left
# axis.ticks.y.right
# axis.ticks.length
# axis.ticks.length.x
# axis.ticks.length.x.top
# axis.ticks.length.x.bottom
# axis.ticks.length.y
# axis.ticks.length.y.left
# axis.ticks.length.y.right
# axis.line
# axis.line.x
# axis.line.x.top
# axis.line.x.bottom
# axis.line.y
# axis.line.y.left
# axis.line.y.right
# legend.background
# legend.margin
# legend.spacing
# legend.spacing.x
# legend.spacing.y
# legend.key
# legend.key.size
# legend.key.height
# legend.key.width
# legend.text
# legend.text.align
# legend.title
# legend.title.align
# legend.position
# legend.direction
# legend.justification
# legend.box
# legend.box.just
# legend.box.margin
# legend.box.background
# legend.box.spacing
# panel.background
# panel.border
# panel.spacing
# panel.spacing.x
# panel.spacing.y
# panel.grid
# panel.grid.major
# panel.grid.minor
# panel.grid.major.x
# panel.grid.major.y
# panel.grid.minor.x
# panel.grid.minor.y
# panel.ontop
# plot.background
# plot.title
# plot.title.position
# plot.subtitle
# plot.caption
# plot.caption.position
# plot.tag
# plot.tag.position
# plot.margin
# strip.background
# strip.background.x
# strip.background.y
# strip.clip
# strip.placement
# strip.text
# strip.text.x
# strip.text.x.bottom
# strip.text.x.top
# strip.text.y
# strip.text.y.left
# strip.text.y.right
# strip.switch.pad.grid
# strip.switch.pad.wrap
```

</div>

### Examples

Below, we customize various elements of a graph (with varying levels of aesthetic appeal...). You will notice that it is quite intuitive to determine whether to use `element_text()`, `element_line()`, or `element_rect()` depending on the context. Note that the argument names are consistent across these three functions (color, size...), which makes them easy to use.

```{r prepelisaraster}
url <- "https://zenodo.org/record/8210893/files/elisa_artificial.txt"
elisa <- read.table(url, sep="\t", header=TRUE, row.names=1)

elisa$rows <- factor(x = elisa$rows, ordered = T, levels=c('cont', letters[1:7]))
p <- ggplot(data = elisa, 
            mapping = aes(x=columns, 
                          y=rows, 
                          fill=value))
p <- p + geom_raster() + 
          facet_grid(facets = user ~ day) +
          scale_fill_gradientn(colours = c("#0000BF", "#0000FF", 
                                           "#0080FF", "#00FFFF", 
                                           "#40FFBF", "#80FF80", 
                                           "#BFFF40", "#FFFF00", 
                                           "#FF8000", "#FF0000", 
                                           "#BF0000"))
print(p)
```

```{r themeexample, exercise=TRUE, exercise.setup="prepelisaraster" }
p <- p + theme_minimal()
p <- p + theme(strip.background = element_rect(color="red", fill="orange"),  
               strip.text       = element_text(color="white", face="bold"),
               axis.text.x      = element_text(color="blue", size=7, angle=45, family = "Helvetica", face="bold"),
               axis.text.y      = element_text(color="darkviolet", size=10, family = "Times", face="bold"),
               axis.ticks.x     = element_line(color="brown", linewidth=1),
               axis.ticks.y     = element_line(color="darkturquoise", linewidth=1),
               plot.background  = element_rect(fill="paleturquoise"),  
              )
#...
```

### Exercises

<div class="exo">

In the following plot:

- Change the font of the graph title (`family="Times"`).
- Adjust the angle of the x-axis text (angle=45°).
- Modify the background color (*fill*).
- Remove the secondary grid lines (using *element_blank()*).
- Add a border line to the boxes containing the legends.

```{r themingexo1, exercise=TRUE, exercise.setup="preppenguins" }
p <- p + ggtitle("Flipper Lengths vs Bill Lengths") +
         theme(plot.title = ___, 
               axis.text.x = ___,
               plot.background = ___, 
               panel.grid.minor = ___,  
               legend.background = element_blank(),
               legend.box.background = ___
           )
print(p)
```

```{r themingexo1-solution}
p <- p + ggtitle("Flipper Lengths vs Bill Lengths") +
               theme(plot.title  = element_text(family="Times"), 
               axis.text.x = element_text(angle=45),
               plot.background = element_rect(fill="#EEDDAA"), 
               panel.grid.minor = element_blank(),  
               legend.background = element_blank(),
               legend.box.background = element_rect(color = "black", size=1)
           )
print(p)
```

</div>

## Exercises

### The Dataset

- Here, our dataset contains several pieces of information related to nearly all known transcripts in the human genome (one per row). This data was produced in *tsv* format using the *pygtftk* software (v1.6.3) from a GTF file downloaded from Ensembl (*genome version GRCh38, release 92*).

Since the file is somewhat large, we will download it and place it in your user folder so that it does not need to be downloaded again later.

```{r prepare_tx_human, exercise=TRUE}
options(timeout=10000)
dir_path <- file.path(fs::path_home(), ".rtrainer")
dir.create(dir_path, showWarnings = FALSE)
## The URL pointing to the dataset
url <- "https://zenodo.org/record/8211383/files/Homo_sapiens.GRCh38.110.subset_2.tsv.gz"
# Download
file_path <- file.path(dir_path, "Homo_sapiens.GRCh38.110.subset_2.tsv.gz")
if(!file.exists(file_path)) download.file(url=url, destfile = file_path, quiet = TRUE)
```

We will load the file into R using the `read.table()` function. At the same time, we will assign the `transcript_id` column to the row names (`row.names=6`).

```{r load_tx_info, exercise=TRUE, exercise.setup="prepare_tx_human" }
tx_info <- read.table(file=file_path, header=TRUE, sep="\t", row.names=6)
dim(tx_info)
```

Here is our dataset:

```{r head_tx_info, exercise=TRUE, exercise.setup="load_tx_info" }
head(tx_info)
```

### Number of Transcripts per Chromosome

<div class="exo">

Create a diagram with `geom_bar()` showing the number of different transcripts per chromosome (`seqid`). 
- Use ` + coord_flip()` to rotate the diagram.
- Order the chromosomes as follows: 1, 2, 3 .. 22, X, Y, MT.

```{r nbgenes, exercise=TRUE , exercise.setup="load_tx_info"}
___
p <- ggplot(data=tx_info, ___) +
  ___
```

```{r nbgenes-solution}
tx_info$seqid <- factor(tx_info$seqid, 
                        levels = c(as.character(1:22), "X", "Y", "MT"), 
                        ordered = TRUE)
p <- ggplot(data=tx_info, 
            mapping=aes(x=seqid)) + 
  geom_bar() + coord_flip() +
  theme_bw()
print(p)
```

```{r nbgenes-check }

gradethis::grade_result_strict(
  pass_if(~ as.character(p$mapping$x)[2] == "seqid"),
  pass_if(~ p$labels$y == "count"),
  pass_if(~ class(p$layers[[1]]$geom)[1] == "GeomBar")
)
```
</div>

### Transcript Sizes

<div class="exo">

- In the `data.frame` `tx_info`, create a new column `tx_genomic_size_log10` containing the variable `tx_genomic_size` converted to base 10 logarithm (`log10()`). Using histograms and facets, explore the variable `tx_genomic_size_log10` (transcript size including introns in base 10 logarithm).

- Use `geom_histogram()` and `facet_grid(gene_biotype~., scale="free_y")`. The argument `scale="free_y"` allows each facet to have its own specific scale.
- Appropriately configure (`theme()`) the size and orientation of the textual elements.

 
```{r txgenomicsizeprep, echo=FALSE}
options(timeout=10000)
dir_path <- file.path(fs::path_home(), ".rtrainer")
dir.create(dir_path, showWarnings = FALSE)
## The URL
url <- "https://zenodo.org/record/8211383/files/Homo_sapiens.GRCh38.110.subset_2.tsv.gz"
# Downloading 
file_path <- file.path(dir_path, "Homo_sapiens.GRCh38.110.subset_2.tsv.gz")
if(!file.exists(file_path)) download.file(url=url, destfile = file_path, quiet = TRUE)
tx_info <- read.table(file=file_path, header=TRUE, sep="\t", row.names=6)
tx_info$tx_genomic_size_log10 <- log10(tx_info$tx_genomic_size)

```

```{r displaytxgenomicsize, echo=FALSE, fig.height=6, fig.width=4 }
options(timeout=10000)
dir_path <- file.path(fs::path_home(), ".rtrainer")
dir.create(dir_path, showWarnings = FALSE)
## The URL
url <- "https://zenodo.org/record/8211383/files/Homo_sapiens.GRCh38.110.subset_2.tsv.gz"
# Downloading 
file_path <- file.path(dir_path, "Homo_sapiens.GRCh38.110.subset_2.tsv.gz")
if(!file.exists(file_path)) download.file(url=url, destfile = file_path, quiet = TRUE)
tx_info <- read.table(file=file_path, header=TRUE, sep="\t", row.names=6)
tx_info$tx_genomic_size_log10 <- log10(tx_info$tx_genomic_size)
p <- ggplot(data=tx_info, 
            mapping=aes(x=tx_genomic_size_log10)) + 
     geom_histogram(bins=100) +
     facet_grid(gene_biotype~., scale="free_y") + 
     labs(x="Genomic Size of Transcripts (log10)") +
     theme_bw() +
     theme(panel.grid.minor = element_blank(), 
           strip.text.y = element_text(angle=0, size=5), 
           axis.text.y = element_text(size=5))
print(p)
```


```{r txgenomicsize, exercise=TRUE , exercise.setup="txgenomicsizeprep"}

```

```{r txgenomicsize-solution}
p <- ggplot(data=tx_info, 
            mapping=aes(x=tx_genomic_size_log10)) + 
     geom_histogram(bins=50) +
     facet_grid(gene_biotype~., scale="free_y") + 
     labs(x="Genomic Size of Transcripts (log10)") +
     theme_minimal() +
     theme(panel.grid.minor = element_blank(), 
           strip.text.y = element_text(angle=0, size=5), 
           axis.text.y = element_text(size=5))
     
```

```{r txgenomicsize-check }

gradethis::grade_result_strict(
  pass_if(~ length(levels(ggplot2::ggplot_build(p)$data[[1]]$PANEL)) == 12),
  pass_if(~ p$labels$y == "count"),
  pass_if(~ class(p$layers[[1]]$geom)[1] == "GeomBar")
)
```


```{r q1, echo=FALSE }
learnr::question("Based on the source available at [this address](http://www.ensembl.org/info/genome/genebuild/biotypes.html), how can we explain that the size of the *processed pseudogene* class is smaller than the *protein coding* class?",
  answer("It is thought that these genes lack UTRs."),
  answer("It is thought that these genes have shorter exons."),
  answer("It is thought that they result from a mature RNA (therefore without introns) reinserting itself, after reverse transcription, into the genome.", correct = TRUE),
  allow_retry = TRUE 
)
```

</div> 

### Number of Exons

If *processed pseudogenes* no longer have introns, we should only find one exon...

<div class="exo">

Transform the `nb_exons` column into its logarithm and place the result in the `nb_exons_log10` column. What can you say about the number of exons (`nb_exons_log10`) for transcripts based on "gene_biotype"? Use a *boxplot* or *violin plot* to present this information.

```{r nbexons, exercise=TRUE , exercise.setup="txgenomicsizeprep"}

```

```{r nbexons-solution}
biotypes <- unique(tx_info$gene_biotype)
palette <- setNames(rainbow(length(biotypes)), biotypes)
tx_info$nb_exons_log10 <- log10(tx_info$nb_exons)
p <- ggplot(data=tx_info, 
            mapping=aes(x=gene_biotype, 
                        y=nb_exons_log10,
                        fill=gene_biotype)) + 
     geom_boxplot() +
     theme_minimal() + 
     labs(y="Nummber of exons (log10)") +
     coord_flip() +
     scale_fill_manual(values=palette) +
     theme(legend.position = "none")
     
print(p)
```

```{r nbexons-check }
gradethis::grade_result_strict(
  pass_if(~ as.character(p$mapping$y)[2] == "nb_exons_log10"),
  pass_if(~ as.character(p$mapping$x)[2] == "gene_biotype"),
  pass_if(~ class(p$layers[[1]]$geom)[1] == "GeomBoxplot")
  
)
```
</div>

### Chromosomal Distribution of Gene Types

<div class="exo">

- Create a bar chart (`geom_bar`) showing the number of transcripts for each `gene_biotype` class on each chromosome. Use `geom_bar()` with the `position` argument set to `stack`, `dodge`, or `fill`. Depending on this argument, do you get the same impression about the distribution of gene biotypes across chromosomes? What are the advantages and disadvantages of each representation? What can you say about the number and types of genes present on the Y chromosome?


```{r distchrombiotypes, exercise=TRUE , exercise.setup="txgenomicsizeprep" }

```

```{r distchrombiotypes-solution}

tx_info$nb_exons_log10 <- log10(tx_info$nb_exons)
p <- ggplot(data=tx_info, 
            mapping=aes(x=seqid, 
                        fill=gene_biotype)) + 
     geom_bar(position="fill") +
     theme_minimal() + 
     labs(y="Count", 
          x="Chromosome") +
     coord_flip() +
     theme(legend.position = "bottom")
     
print(p)
```

```{r distchrombiotypes-check }
gradethis::grade_result_strict(
  pass_if(~ as.character(p$mapping$x)[2] == "seqid"),
  pass_if(~ as.character(p$mapping$fill)[2] == "gene_biotype"),
  pass_if(~ class(p$layers[[1]]$geom)[1] == "GeomBar")
  
)
```
</div>

### End of the section

Thank you for following this tutorial.
