---
# theme
# “bootstrap”, “cerulean”, “cosmo”, “darkly”, “flatly”, “journal”, “lumen”, “paper”, 
# “readable”, “sandstone”, “simplex”, “spacelab”, “united”, “yeti”
# highlight: `default`, `tango`, `pygments`, `kate`, `monochrome`, `espresso`, `zenburn`, 
# `haddock`, `breezedark`, `textmate`, `arrow`, or `rstudio` or a file with extension `.theme`.

title: "Object-Oriented Programming in R"
output:
  learnr::tutorial:
    includes:
      before_body: !expr system.file(file.path("tutorials", "style.html"),package="rtrainer")
    theme: default
    highlight: default
    fig_caption: yes
    self_contained: true
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 4
    number_sections: false
    progressive: true
  html_document:
    theme: cosmo
    fig_caption: yes
    self_contained: yes
    toc: yes
    toc_float: 
      toc_collapsed: false
    toc_depth: 3
    number_section: true
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
runtime: shiny_prerendered
---



<!--  
Here the parameters about the documents.
https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf 
--> 

```{css, echo=FALSE}

```

<script language="JavaScript" type="text/javascript">
          
          function sizeTbl2(h,i) {
          var tbl = document.getElementById("section-" + i);
          tbl.style.display = h;
          }

</script>



<style>
.exo {
  border-radius: 5px;
  margin-top: 5px;
  margin-bottom: 5px;
  padding-top: 5px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  background-color: #fcede3;
  color: rgb(51, 51,153);
}
.tips {
       padding-top: 5px;
       padding-bottom: 5px;
       padding-left: 5px;
       padding-right: 5px;
       border: 1px dashed #2f6fab;
       background-color: #EEFFEE;
}
.solution {
            margin-top: 5px;
            margin-bottom: 5px;
            padding-top: 5px;
            padding-bottom: 5px;
            padding-left: 5px;
            padding-right: 5px;
            border: 1px dashed #FFFFFF;
            background-color: #EEEEFF;
            color: #0000BB;
            font-size: 11px;
}
</style>


```{r setup, include=FALSE}
library(learnr)
library(gradethis)
knitr::opts_chunk$set(echo=TRUE, eval=TRUE, cache=FALSE, message=FALSE, warning=FALSE, comment="")
gradethis::gradethis_setup()
```


## Introduction to Class Systems in R

**Class systems** are a fundamental element of **object-oriented programming in R**. They allow the creation of complex data structures called objects, which combine **attributes (variables)** and **methods (functions)** to **represent real-world entities in a modular way**. In R, the two main class systems are **S3 and S4**, each with its own characteristics and conventions. In this course, we will focus on S4 classes, which are commonly used in the **Bioconductor** project, a set of function libraries for biological data analysis.

### S4 Classes

#### Creating a 'single_cell' Class

Below is an example of how to implement a `single_cell` class that allows for storing a gene expression matrix, metadata about the cells (a `data.frame`), gene information (a `data.frame`), and a list containing free-form information about the experiment:

```{r single_cell, exercise=TRUE}
# Définir la classe S4 "single_cell"
setClass(
  Class = "single_cell",
  representation(
    expression_matrix = "matrix",
    cell_metadata = "data.frame",
    gene_information = "data.frame",
    exp_info="list"
  )
)
```

#### Definition of the Constructor Method

In object-oriented programming (OOP), **a method** is a **function associated with a class**. It defines the **behavior or actions that objects of this class can perform**. Methods are called on objects to perform specific operations.

The "initialize" method is a special method in R, often called a **constructor method**. It is used to **initialize an object of the class when it is created**. More specifically, its purpose is to **configure or assign initial values to the attributes of the object**, so that the object is ready to be used in your application. In the case of the "single_cell" class you have defined, the `initialize()` method is responsible for initializing the attributes of the class.

**NB**: The `initialize()` method expects an argument named `.Object` (*see* section *Usage* `?summary`).


```{r initialize, exercise=TRUE, exercise.setup="single_cell", exercise.lines=20}
  # Create a constructor for the "single_cell" class
setMethod("initialize", 
          signature(.Object = "single_cell"),
  function(.Object, 
           expression_matrix=matrix(), 
           cell_metadata=data.frame(), 
           gene_information=data.frame(),
           exp_info=list(author=NULL, date=NULL, laboratory=NULL)) {
    
    .Object@expression_matrix <- expression_matrix
     colnames(.Object@expression_matrix) <- cell_data$cell_id
     rownames(.Object@expression_matrix) <- gene_information$cell_id
    .Object@cell_metadata <- cell_metadata
    .Object@gene_information <- gene_information
    .Object@exp_info <- exp_info
    
    return(.Object)
  }
)
```

## Creating an Instance

### Creating an Instance of the 'single_cell' Class

Given the following data (matrix `m`, data.frame `cell_data`, data.frame `g_info`):

```{r scdata, exercise=TRUE, exercise.setup="initialize", exercise.lines=16}
set.seed(123)
m <- matrix(rnorm(200), 
            nrow = 20, 
            ncol = 10)

cell_data <- data.frame(cell_id = 1:10, 
                        cell_pop = 1:10, 
                        sample_type = sample(c("A", "B"), 10, replace=TRUE))

g_info <- data.frame(gene_id = 1:20, 
                     gene_name = letters[1:20])

exp_info <- list(author="D. Puthier", 
                 date="Tue Nov  7 09:29:10 2023", 
                 laboratory="TAGC")
```

We will use the following code to store them in an object of the 'single_cell' class. We say that the object `sc_data` is **an instance of the 'single_cell' class**.


```{r new, exercise=TRUE, exercise.setup="scdata"}
sc_data <- new("single_cell",
  expression_matrix = m,
  cell_metadata = cell_data,
  gene_information = g_info,
  exp_info=exp_info
)
```

For now, when we call the object, it shows us everything it contains. If the data is large, this is not practical. Therefore, we will redefine the `show()` method.

```{r showme, exercise=TRUE, exercise.setup="new"}
sc_data
```

## Defining Generic Methods

### Defining the 'show()' Method

A number of methods are called generic. These methods exist by default in the R language (e.g., `ncol()`, `nrow()`, `dim()`, `summary()`, `show()`, `[`...) and are defined for many different objects. Similarly, we can redefine them for our 'single_cell' object.
The `show()` method is the so-called "representation" method. It allows us to define a display message when an object of the 'single_cell' class is called (*i.e.*, when its name is typed in the terminal and Enter is pressed). This function exists for all objects, but we are redefining it here for our 'single_cell' object.

**NB**: By default, this function expects an argument named `object` (*see* `?show`).

```{r show, exercise=TRUE, exercise.setup="new"}

# Let's define the `show()`
setMethod("show", 
          signature(object = "single_cell"),
  function(object) {
    
    nr <- nrow(object@expression_matrix)
    nc <- ncol(object@expression_matrix)
    
    cat("A single cell object with:\n")
    cat(paste0("\t- ", nr, " genes\n"))
     cat(paste0("\t- ", nc, " samples"))  
  }
)

```

When this object is called, it is displayed via its `show()` method:

```{r showcall, exercise=TRUE, exercise.setup="show"}
sc_data
```

### Defining ncol() and nrow()

<div class="exo">

Define the two methods `ncol()` and `nrow()` that return the number of rows and columns, respectively, of the `expression_matrix` from a 'single_cell' object. You can refer to the code provided below:

**NB**: `ncol()` and `nrow()` are generic functions where the only argument is `x` (*see* `?ncol`).

```{r ncolnrow, exercise=TRUE, exercise.setup="show", exercise.lines=15}
setMethod("ncol", 
          signature(x = "single_cell"),
  function(x) {

  }
)
```

```{r ncolnrow-solution}
setMethod("ncol", 
          signature(x = "single_cell"),
  function(x) {
    return(ncol(x@expression_matrix))
  }
)

setMethod("nrow", 
          signature(x = "single_cell"),
  function(x) {
    return(nrow(x@expression_matrix))
  }
)
```

```{r ncolnrow-check}

gradethis::grade_result_strict(
  pass_if(~ nrow(sc_data) == 20),
  pass_if(~ ncol(sc_data) == 10)
)
```
</div>

### Defining summary()

<div class="exo">

Define the generic method `summary()` that displays a numeric summary of an object. The `summary()` method expects 'object' as an argument (`?summary`). Here, the method should return the mean of the columns of the 'expression_matrix' attribute.

```{r summary, exercise=TRUE, exercise.setup="show", exercise.lines=15}

```

```{r summary-solution }
setMethod("summary", 
          signature(object = "single_cell"),
  function(object) {
    col_means <- colMeans(object@expression_matrix)
    names(col_means) <- colnames(object@expression_matrix)
    return(col_means)
  }
)
```

```{r summary-check}
gradethis::grade_result_strict(
  pass_if(~ nrow(sc_data) == 20),
  pass_if(~ ncol(sc_data) == 10)
)
```
</div>


## Creating New Generic Methods

### Generic Method 'author()'

It is possible to create new, user-defined methods. In this case, you first need to create a generic method and then define it for the targeted object (here, 'single_cell'). For example, here we create the `author()` function, which returns the name of the experimenter (if it is defined).

```{r newgeneric, exercise=TRUE, exercise.setup="show", exercise.lines=15}
setGeneric("author", function(object) {
  standardGeneric("author")
})

setMethod("author", 
          signature(object = "single_cell"),
  function(object) {
    if("author" %in% names(object@exp_info)){
      return(object@exp_info$author)
    }else{
      return(NULL)
    }
  }
)
```

### Generic Method 'gene_name()'

<div class="exo">

Define a new function `gene_name()` that returns the list of gene names (`gene_name`) from the 'gene_information' attribute.

```{r gene_name, exercise=TRUE, exercise.setup="show", exercise.lines=15}

```

```{r gene_name-solution}
setGeneric("gene_name", function(object) {
  standardGeneric("gene_name")
})

setMethod("gene_name", 
          signature(object = "single_cell"),
  function(object) {
    return(object@gene_information$gene_name)
  }
)
```

```{r gene_name-check}
gradethis::grade_result_strict(
  pass_if(~ all(gene_name(sc_data) == letters[1:20]) == TRUE)
)
```
</div>

## Summary

### Complete Code for the Class

To summarize, the following code allows us to create a `single_cell` class and its associated methods.

```{r summarize_all, exercise=TRUE, exercise.lines=90}
# Définir la classe S4 "single_cell"
setClass(
  Class = "single_cell",
  representation(
    expression_matrix = "matrix",
    cell_metadata = "data.frame",
    gene_information = "data.frame",
    exp_info="list"
  )
)

# Create a constructor for the "single_cell" class
setMethod("initialize", 
          signature(.Object = "single_cell"),
  function(.Object, 
           expression_matrix=matrix(), 
           cell_metadata=data.frame(), 
           gene_information=data.frame(),
           exp_info=list(author=NULL, date=NULL, laboratory=NULL)) {
    
    .Object@expression_matrix <- expression_matrix
     colnames(.Object@expression_matrix) <- cell_data$cell_id
     rownames(.Object@expression_matrix) <- gene_information$cell_id
    .Object@cell_metadata <- cell_metadata
    .Object@gene_information <- gene_information
    .Object@exp_info <- exp_info
    
    return(.Object)
  }
)

# Define the generic method `show()`
setMethod("show", 
          signature(object = "single_cell"),
  function(object) {
    
    nr <- nrow(object@expression_matrix)
    nc <- ncol(object@expression_matrix)
    
    cat("A single cell object with:\n")
    cat(paste0("\t- ", nr, " genes\n"))
     cat(paste0("\t- ", nc, " samples"))  
  }
)

setMethod("ncol", 
          signature(x = "single_cell"),
  function(x) {
    return(ncol(x@expression_matrix))
  }
)

setMethod("nrow", 
          signature(x = "single_cell"),
  function(x) {
    return(nrow(x@expression_matrix))
  }
)

setGeneric("author", function(object) {
  standardGeneric("author")
})

setMethod("author", 
          signature(object = "single_cell"),
  function(object) {
    if("author" %in% names(object@exp_info)){
      return(object@exp_info$author)
    }else{
      return(NULL)
    }
  }
)


setGeneric("gene_name", function(object) {
  standardGeneric("gene_name")
})

setMethod("gene_name", 
          signature(object = "single_cell"),
  function(object) {
    return(object@gene_information$gene_name)
  }
)

```

## End of the section

Thank you for following this tutorial.

